### PR TITLE
fix well iterations counting bugs.

### PR DESCRIPTION
**Pull request by qilicun**
If the well iterations is ``std::numeric_limits<int>::min()``, it should not do the counting.

### CODE REVIEW

File: opm/autodiff/NonlinearSolver_impl.hpp [hunk 1]
```diff
@@ -137,7 +137,11 @@ namespace Opm
             }
             converged = report.converged;
             linIters += report.linear_iterations;
-            wellIters += report.well_iterations;
+            if (report.well_iterations != std::numeric_limits<int>::min()) {
+                wellIters += report.well_iterations;
+            } else {
+                wellIters = report.well_iterations;
```
Comment (bska) [inline opm/autodiff/NonlinearSolver_impl.hpp]: Are you sure about this?  Isn't this branch (i.e., the `else` branch) exactly equal to
```C++
wellIters = std::numeric_limits<int>::min();
```
?  Won't that lead to problems elsewhere?

Comment (qilicun) [inline opm/autodiff/NonlinearSolver_impl.hpp]: ``report.well_iterations`` is initialized with ``std::numeric_limits<int>`` in all the place, if ``solveWellEq()`` is not called, it will stay the same, I think.

Comment (bska) [inline opm/autodiff/NonlinearSolver_impl.hpp]: > if `solveWellEq()` is not called, it will stay the same, I think.

Sure, but that's not the point I was trying to make.  A few [lines down](https://github.com/qilicun/opm-simulators/blob/bb0164f39f2979549dbac75b66622a24ed71260b/opm/autodiff/NonlinearSolver_impl.hpp#L157) from this loop we find the statement
```C++
wellIterations_ += wellIters; 
```
This is undefined behaviour (integer overflow) when both of these conditions are true

* `wellIters == std::numeric_limits<int>::min()`
*  `wellIterations_ < 0`

In other words, if we assign `wellIters = std::numeric_limits<int>::min()` as this patch does in the `else` branch then we need to take special care in the `wellIters_` update lest the program be undefined.

Comment (qilicun) [inline opm/autodiff/NonlinearSolver_impl.hpp]: Sorry for not noticing that, ``wellIterations_`` should be taken care as well.

**Comment by atgeirr** [comment]
I'm getting a little cold feet about the use of numeric_limits::min for this purpose, it looks like it would be much simpler to understand and maintain if we just return a 0 if the feature is not used? We would not be able to distinguish between the feature not being used, and it being used but already converged with no iterations, but is that a problem? (We'd have to change the printout to only print if number of iterations > 0.)

**Comment by qilicun** [comment]
> I'm getting a little cold feet about the use of numeric_limits::min for this purpose, it looks like it would be much simpler to understand and maintain if we just return a 0 if the feature is not used?

I agree, PR is updated, will make a relative PR to opm-core.

File: opm/autodiff/NonlinearSolver_impl.hpp [hunk 2]
```diff
@@ -150,10 +150,10 @@ namespace Opm
 
         linearIterations_ += linIters;
         nonlinearIterations_ += iteration - 1; // Since the last one will always be trivial.
-        wellIterations_ += wellIters; 
+        wellIterations_ = wellIters;
```
Comment (bska) [inline opm/autodiff/NonlinearSolver_impl.hpp]: Should this be `+=` (update) rather than `=` (assignment)?

Comment (qilicun) [inline opm/autodiff/NonlinearSolver_impl.hpp]: I think it should be ``=``, because every sub time step, well equation will be solved or not, if not, this ``+=`` will give the same number of last sub time step, I do not think that makes sense.

Comment (bska) [inline opm/autodiff/NonlinearSolver_impl.hpp]: > I think it should be `=`, because every sub time step, well equation will be solved or not, if not, this `+=` will give the same number of last sub time step, I do not think that makes sense.

Then we can retire one of `wellIterations_` and `wellIterationsLast_` because these quantites now have the exact same semantics.  The `wellIterations_` counter **used** to mean total number of well iterations for the entire simulation and now it does not.

Comment (qilicun) [inline opm/autodiff/NonlinearSolver_impl.hpp]: Removed.

**Comment by bska** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by bska** [comment]
I've decided that I really want the original `wellIterations_` behaviour because we'd otherwise be removing useful information in the common case of active wells.  Please restore the `wellIterations_` field, including the `+=` update at the end of `step()` and then I'll merge.

**Comment by bska** [comment]
Thanks.  I'll merge into master.
